tests/lib/rules/no-prototype-builtins.js

```diff
@@ -153,22 +201,95 @@ ruleTester.run("no-prototype-builtins", rule, {
         {
             code: "foo?.hasOwnProperty('bar')",
             parserOptions: { ecmaVersion: 2020 },
-            errors: [{ messageId: "prototypeBuildIn", data: { prop: "hasOwnProperty" } }]
+            errors: [{ messageId: "prototypeBuildIn", data: { prop: "hasOwnProperty" }, suggestions: [] }]
+        },
+        {
+            code: "foo?.bar.hasOwnProperty('baz')",
+            parserOptions: { ecmaVersion: 2020 },
+            errors: [{ messageId: "prototypeBuildIn", data: { prop: "hasOwnProperty" }, suggestions: [] }]
+        },
+        {
+            code: "foo.hasOwnProperty?.('bar')",
+            parserOptions: { ecmaVersion: 2020 },
+            errors: [{ messageId: "prototypeBuildIn", data: { prop: "hasOwnProperty" }, suggestions: [] }]
+        },
+        {
+
+            /*
+             * if hasOwnProperty is part of a ChainExpresion
+             * and the optional part is before it, then don't suggest the fix
+             */
+            code: "foo?.hasOwnProperty('bar').baz",
+            parserOptions: { ecmaVersion: 2020 },
+            errors: [{ messageId: "prototypeBuildIn", data: { prop: "hasOwnProperty" }, suggestions: [] }]
+        },
+        {
+
+            /*
+             * if hasOwnProperty is part of a ChainExpresion
+             * but the optional part is after it, then the fix is safe
+             */
+            code: "foo.hasOwnProperty('bar')?.baz",
+            parserOptions: { ecmaVersion: 2020 },
+            errors: [{
+                messageId: "prototypeBuildIn",
+                data: { prop: "hasOwnProperty" },
+                suggestions: [
+                    {
+                        messageId: "callObjectPrototype",
+                        output: "Object.prototype.hasOwnProperty.call(foo, 'bar')?.baz"
+                    }
+                ]
+            }]
         },
         {
             code: "(foo?.hasOwnProperty)('bar')",
             parserOptions: { ecmaVersion: 2020 },
-            errors: [{ messageId: "prototypeBuildIn", data: { prop: "hasOwnProperty" } }]
+            errors: [{
+                messageId: "prototypeBuildIn",
+                data: { prop: "hasOwnProperty" },
+                suggestions: [
+                    {
+
+                        /*
+                         * Note that the original may throw TypeError: (intermediate value) is not a function
+                         * whereas the replacement may throw TypeError: Cannot convert undefined or null to object
+                         */
+                        messageId: "callObjectPrototype",
+                        output: "(Object.prototype.hasOwnProperty.call)(foo, 'bar')"
+                    }
+                ]
+            }]
+        },
```

Comment:
I think it would be better not to provide the suggestion in cases like this (`node.callee` is a ChainExpression).

This is a possible error reported by [no-unsafe-optional-chaining](https://eslint.org/docs/latest/rules/no-unsafe-optional-chaining) and should be fixed manually.

Reply:
Thanks I stopped giving the suggestion when `node.callee.type === "ChainExpression"`. I do wonder whether we should remove the `const callee = astUtils.skipChainExpression(node.callee);` above and just not report any no-prototype-builtins for `(foo?.hasOwnProperty)('bar')` in favor of no-unsafe-optional-chaining.

Reply:
> I do wonder whether we should remove the `const callee = astUtils.skipChainExpression(node.callee);` above and just not report any no-prototype-builtins for `(foo?.hasOwnProperty)('bar')` in favor of no-unsafe-optional-chaining.

I think it's fine for both rules to report this code as those are two different problems.